File: backend/matching-engine/src/test/java/com/commonsengine/matching/strategy/NearestFirstStrategyTest.java
package com.commonsengine.matching.strategy;

import com.commonsengine.platform.domain.Model.ConsumerId;
import com.commonsengine.platform.domain.Model.MatchResult;
import com.commonsengine.platform.domain.Model.RequestId;
import com.commonsengine.platform.domain.Model.ServiceRequest;
import com.commonsengine.platform.domain.Model.Worker;
import com.commonsengine.platform.domain.Model.WorkerId;
import com.commonsengine.platform.domain.ServiceType;
import com.commonsengine.platform.geo.GeoPoint;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * 最近优先策略测试——从 Kotlin NearestFirstStrategyTest 转换。
 *
 * <p>适配 Java API：
 * <ul>
 *   <li>{@code NearestFirstStrategy()} / {@code NearestFirstStrategy(AntiExploitationConfig)}</li>
 *   <li>{@code match(request, candidates)} → MatchResult(workerId, requestId, score)</li>
 *   <li>AntiExploitationConfig 字段：{@code maxActiveOrdersPerWorker}（默认 3）</li>
 * </ul>
 */
class NearestFirstStrategyTest {

    private final GeoPoint pickup = new GeoPoint(39.9042, 116.4074); // 北京天安门

    private ServiceRequest request() {
        return new ServiceRequest(
                RequestId.random(),
                new ConsumerId("c1"),
                ServiceType.RIDE_HAILING,
                pickup,
                new GeoPoint(39.9300, 116.4300)
        );
    }

    private Worker worker(String id, double lat, double lng) {
        return new Worker(
                new WorkerId(id),
                "worker-" + id,
                new GeoPoint(lat, lng),
                5.0,
                0,
                Set.of()
        );
    }

    private Worker worker(String id, double lat, double lng, int orders, double rating) {
        return new Worker(
                new WorkerId(id),
                "worker-" + id,
                new GeoPoint(lat, lng),
                rating,
                orders,
                Set.of()
        );
    }

    @Test
    void selectsNearestWorker() {
        NearestFirstStrategy strategy = new NearestFirstStrategy();
        List<Worker> candidates = List.of(
                worker("far", 39.9200, 116.4200),   // ~2km
                worker("near", 39.9050, 116.4080),  // ~100m
                worker("mid", 39.9100, 116.4100)    // ~700m
        );

        MatchResult result = strategy.match(request(), candidates);

        assertNotNull(result);
        assertEquals("near", result.workerId().value());
        // score = 1/(1+distance)，最近者 score 最高
        assertTrue(result.score() > 0);
        assertEquals("nearest-first", strategy.name());
    }

    @Test
    void filtersOutOverloadedWorkers() {
        // maxActiveOrdersPerWorker = 2 → 活跃订单数 ≥ 2 的劳动者被排除
        NearestFirstStrategy strategy = new NearestFirstStrategy(
                new AntiExploitationConfig(2, 0, 0)
        );
        List<Worker> candidates = List.of(
                worker("overloaded", 39.9050, 116.4080, 3, 5.0),  // 很近但超载
                worker("available", 39.9100, 116.4100, 0, 5.0)    // 稍远但可用
        );

        MatchResult result = strategy.match(request(), candidates);

        assertNotNull(result);
        assertEquals("available", result.workerId().value());
    }

    @Test
    void returnsNullWhenAllCandidatesOverloaded() {
        NearestFirstStrategy strategy = new NearestFirstStrategy(
                new AntiExploitationConfig(0, 0, 0)
        );
        List<Worker> candidates = List.of(
                worker("a", 39.9050, 116.4080, 1, 5.0)  // activeOrderCount=1 > 0
        );

        MatchResult result = strategy.match(request(), candidates);
        assertNull(result);
    }

    @Test
    void emptyCandidateListReturnsNull() {
        NearestFirstStrategy strategy = new NearestFirstStrategy();
        MatchResult result = strategy.match(request(), List.of());
        assertNull(result);
    }
}

File: backend/matching-engine/src/test/kotlin/com/commonsengine/matching/strategy/NearestFirstStrategyTest.kt
package com.commonsengine.matching.strategy

import com.commonsengine.platform.domain.ConsumerId
import com.commonsengine.platform.domain.RequestId
import com.commonsengine.platform.domain.ServiceRequest
import com.commonsengine.platform.domain.ServiceType
import com.commonsengine.platform.domain.Worker
import com.commonsengine.platform.domain.WorkerId
import com.commonsengine.platform.geo.GeoPoint
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

class NearestFirstStrategyTest {

    private val pickup = GeoPoint(39.9042, 116.4074) // 北京天安门

    private fun request() = ServiceRequest(
        id = RequestId.random(),
        consumerId = ConsumerId("c1"),
        type = ServiceType.RIDE_HAILING,
        pickupLocation = pickup,
    )

    private fun worker(id: String, lat: Double, lng: Double, orders: Int = 0, rating: Double = 5.0) = Worker(
        id = WorkerId(id),
        name = "worker-$id",
        currentLocation = GeoPoint(lat, lng),
        rating = rating,
        activeOrderCount = orders,
    )

    @Test
    fun `selects nearest worker`() {
        val strategy = NearestFirstStrategy()
        val candidates = listOf(
            worker("far", 39.9200, 116.4200),   // ~2km
            worker("near", 39.9050, 116.4080),  // ~100m
            worker("mid", 39.9100, 116.4100),   // ~700m
        )

        val result = strategy.match(request(), candidates)

        assertNotNull(result)
        assertEquals("near", result.worker.id.value)
        assertTrue(result.distanceMeters < 200, "最近的劳动者应在 200 米内，实际 ${result.distanceMeters}")
        assertTrue(result.reason.contains("距离最近"), "理由应包含'距离最近'")
    }

    @Test
    fun `returns null when no candidates within radius`() {
        val strategy = NearestFirstStrategy(
            AntiExploitationConfig(maxMatchRadiusMeters = 100.0),
        )
        val candidates = listOf(
            worker("far", 40.0000, 116.5000),  // 很远
        )

        val result = strategy.match(request(), candidates)
        assertNull(result)
    }

    @Test
    fun `filters out overloaded workers`() {
        val strategy = NearestFirstStrategy(
            AntiExploitationConfig(maxActiveOrders = 2),
        )
        val candidates = listOf(
            worker("overloaded", 39.9050, 116.4080, orders = 3),  // 很近但超载
            worker("available", 39.9100, 116.4100, orders = 0),   // 稍远但可用
        )

        val result = strategy.match(request(), candidates)

        assertNotNull(result)
        assertEquals("available", result.worker.id.value)
        assertTrue(result.reason.contains("活跃订单数"))
    }

    @Test
    fun `empty candidate list returns null`() {
        val strategy = NearestFirstStrategy()
        val result = strategy.match(request(), emptyList())
        assertNull(result)
    }
}
